sql: add support for create index inside new schema changer

fqazi opened the pull request:
Fixes: #67264 

Previously, create index was not supported inside the
new schema changer. This was inadequate because to
help support transaction schema changes in the future
we need all schema changes inside the new schema changer.
To address this, this patch adds CREATE INDEX inside
the new schema changer.

Release note: None

fqazi committed:
sql: added support for create index inside new schema changer

Fixes: #67264

Previously, create index was not supported inside the
new shcema changer. This was inadequate because, to
help support transaction schema changes in the future
we need all schema changes inside the new schema changer.
To address this, this patch adds CREATE INDEX inside
the new schema changer.

Release note: None

fqazi committed:
sql: implement unique index validation inside the new

Previously, the unique index validation inside the
new schema changer was stubbed out. This was inadequate
because during index addition operations we were not
properly confirming the unique constraints were properly
implemented. To address this, this patch adds support in
the new schema changer for the valdiation logic.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67266)
<!-- Reviewable:end -->

nihalpednekar commented:
LGTM from Bulk perspective

nihalpednekar commented:
@adityamaru brought to my attention that Bulk also owns some parts of `*/backfill.go` so please wait until someone else from Bulk takes a look at this as well.

ajwerner reviewed (commented):
I'm anxious about the ways that this is running into the pain of descriptors vs. elements. 

> Reviewed 19 of 33 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-:-Mg6N5hV-dVSazwkNaxn:byfp3s7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/schemachanger/scbuild/index.go, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6N2UgDSifnmFKpyrA:-Mg6N2UgDSifnmFKpyrB:b-aywio7) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L67)):*
> ```Go
> //    mod(fnv32(colNames[0]::STRING)+fnv32(colNames[1])+...,buckets)
> //
> func makeHashShardComputeExpr(colNames []string, buckets int) *string {
> ```

I wonder if we can share this code?
___
*[pkg/sql/schemachanger/scbuild/index.go, line 191 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6LX2m8m0JBG01TtFv:-Mg6LX2m8m0JBG01TtFw:b-8so11) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L191)):*
> ```Go
> 
> func (b *buildContext) createIndex(ctx context.Context, n *tree.CreateIndex) {
> 	// Look up the table first
> ```

nit: period
___
*[pkg/sql/schemachanger/scbuild/index.go, line 193 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MFr76E1b8mND8v-E:-Mg6MFr76E1b8mND8v-F:b-joxsr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L193)):*
> ```
> 	_, table, err := resolver.ResolveExistingTableObject(ctx, b.Res, &n.Table,
> 		tree.ObjectLookupFlagsWithRequired())
> ```
we ought to be able to plumb through a mutable flag here to not need to resolve it again. I'd like to know if that does not work. 
___
*[pkg/sql/schemachanger/scbuild/index.go, line 197 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6LW7Z8DOKLy3deUXq:-Mg6LW7Z8DOKLy3deUXr:b-8so11) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L197)):*
> ```Go
> 		panic(err)
> 	}
> 	// Detect if the index already exists
> ```

nit: period
___
*[pkg/sql/schemachanger/scbuild/index.go, line 202 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MT9E0AWnCPiNSwDQ:-Mg6MT9E0AWnCPiNSwDR:bj8854f) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L202)):*
> ```Go
> 		if foundIndex.Dropped() {
> 			panic(pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
> 				"index %q being dropped, try again later", string(n.Name)))
> ```

nit: use the format/string method on the AST node for an index name.
___
*[pkg/sql/schemachanger/scbuild/index.go, line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MM6I-enIJb5yYcgP:-Mg6MM6I-enIJb5yYcgQ:b-xm2e0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L211)):*
> ```Go
> 
> 	if table.IsView() && !table.MaterializedView() {
> 		panic(pgerror.Newf(pgcode.WrongObjectType, "%q is not a table or materialized view", table.GetName()))
> ```

&n.Table?
___
*[pkg/sql/schemachanger/scbuild/index.go, line 244 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6LtHR-YWQs1kZIKgz:-Mg6LtHR-YWQs1kZIKh-:b-32w5h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L244)):*
> ```Go
> 	}
> 	colNames := make([]string, 0, len(n.Columns))
> 	// Setup the column ID's
> ```

nit: period and no `'`.
___
*[pkg/sql/schemachanger/scbuild/index.go, line 259 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6Lykq1HJqfyPxNg-h:-Mg6Lykq1HJqfyPxNg-i:brnm71f) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L259)):*
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

> ```
> 			_, typ, _, err := schemaexpr.DequalifyAndValidateExpr(
> 				ctx,
> 				table,
> 				columnNode.Expr,
> 				types.Any,
> 				"index expression",
> 				b.SemaCtx,
> 				tree.VolatilityImmutable,
> 				&n.Table,
> 			)
> ```
</details>

This is not going to work if the column was added in this transaction. At least we should detect that case by searching the elements and then returning a not supported error. 
___
*[pkg/sql/schemachanger/scbuild/index.go, line 264 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MaH-2xWEF5xxjUA9:-Mg6MaH-2xWEF5xxjUAA:b36qhd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L264)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 			mutableDesc, err := b.Descs.GetMutableTableByID(ctx, b.EvalCtx.Txn, table.GetID(), tree.ObjectLookupFlagsWithRequired())
> 			if err != nil {
> 				panic(err)
> 			}
> ```
</details>

let's avoid this lookup. 
___
*[pkg/sql/schemachanger/scbuild/index.go, line 269 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MjUu10KgZkAnHgXu:-Mg6MjUu10KgZkAnHgXv:b-wlpvd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L269)):*
> ```
> 			// Otherwise, create a new virtual column and add it to the table
> 			// descriptor.
> ```
Otherwise relative to what?
___
*[pkg/sql/schemachanger/scbuild/index.go, line 271 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MozgEks1rKuJKhqH:-Mg6MozgEks1rKuJKhqI:b-eewjek) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L271)):*
> ```Go
> 			// descriptor.
> 			colName := tabledesc.GenerateUniqueName("crdb_idx_expr", func(name string) bool {
> 				_, err := mutableDesc.FindColumnWithName(tree.Name(name))
> ```

Doesn't this need to think about newly added columns? Imagine creating more than one index or creating an index with more than one expression. 
___
*[pkg/sql/schemachanger/scbuild/table.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mf_trBKC7LzeoUsU6x4:-Mf_trBKC7LzeoUsU6x5:b-g94uh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/128955ec965cab2785be2435a192e47c6d1764e7/pkg/sql/schemachanger/scbuild/table.go#L253)):*
> ```Go
> 		FamilyName: familyName,
> 	})
> 	/*newPrimaryIdxID := */ b.addOrUpdatePrimaryIndexTargetsForAddColumn(table, colID, col.Name)
> ```

What is this about?
___
*[pkg/sql/lex/BUILD.bazel, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mf_tVHtBzh9yt_FJbGV:-Mf_tVHtBzh9yt_FJbGW:b-s2mfyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/128955ec965cab2785be2435a192e47c6d1764e7/pkg/sql/lex/BUILD.bazel#L7)):*
> ```Starlark
>     srcs = [
>         "encode.go",
>         "keywords.go",
> ```

I don't think you meant to add these. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67266)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/scbuild/index.go, line 193 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MFr76E1b8mND8v-E:-MgCH6sK-ahzR9ifkpx0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L193)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	_, table, err := resolver.ResolveExistingTableObject(ctx, b.Res, &n.Table,
> 		tree.ObjectLookupFlagsWithRequired())
> ```
we ought to be able to plumb through a mutable flag here to not need to resolve it again. I'd like to know if that does not work. 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scbuild/index.go, line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MM6I-enIJb5yYcgP:-MgC4HJw6nZFkK7cIVCl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L211)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

&n.Table?
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scbuild/index.go, line 264 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MaH-2xWEF5xxjUA9:-MgCBm2f1LOFJldcBjWU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L264)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 			mutableDesc, err := b.Descs.GetMutableTableByID(ctx, b.EvalCtx.Txn, table.GetID(), tree.ObjectLookupFlagsWithRequired())
> 			if err != nil {
> 				panic(err)
> 			}
> ```
let's avoid this lookup. 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scbuild/index.go, line 269 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MjUu10KgZkAnHgXu:-MgCGQXS0zY7xX0mcRXo:b-bwgcab) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L269)):*
> create
Your right no otherwise, this will always be done
___
*[pkg/sql/schemachanger/scbuild/index.go, line 271 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mg6MozgEks1rKuJKhqH:-MgCNH_p2Z-41OjHGEgu:b-qpq9eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cd376729ea749b53780f31baad1567703d2361/pkg/sql/schemachanger/scbuild/index.go#L271)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Doesn't this need to think about newly added columns? Imagine creating more than one index or creating an index with more than one expression. 
</blockquote></details>

Yes, we need to sort out those scenarios, as a stop-gap, I'm returning unsupported errors. The query stuff I think will allow us better interfaces for this stuff too.
___
*[pkg/sql/schemachanger/scbuild/table.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mf_trBKC7LzeoUsU6x4:-MgCD6arC3kcckEZBs5P:b-xctcfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/128955ec965cab2785be2435a192e47c6d1764e7/pkg/sql/schemachanger/scbuild/table.go#L253)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What is this about?
</blockquote></details>

We were incorrectly adding a secondaryIndex with the same ID in the unique case if I remember correctly. So, the newPrimaryIdxId parameter has no real use.
___
*[pkg/sql/lex/BUILD.bazel, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mf_tVHtBzh9yt_FJbGV:-MgCBzlZEN4MCPFFhCRA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/128955ec965cab2785be2435a192e47c6d1764e7/pkg/sql/lex/BUILD.bazel#L7)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think you meant to add these. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pbardea commented:
Bulk plumbing changes LGTM -- removing bulk from this PR.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=67266) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=67266) before we can accept your contribution.<br/>**16** out of **17** committers have signed the CLA.<br/><br/>:white_check_mark: rafiss<br/>:white_check_mark: elkmaster<br/>:white_check_mark: rail<br/>:white_check_mark: rickystewart<br/>:white_check_mark: JeffSwenson<br/>:white_check_mark: otan<br/>:white_check_mark: dhartunian<br/>:white_check_mark: nvanbenschoten<br/>:white_check_mark: yuzefovich<br/>:white_check_mark: fqazi<br/>:white_check_mark: alan-mas<br/>:white_check_mark: ajwerner<br/>:white_check_mark: stevendanna<br/>:white_check_mark: DuskEagle<br/>:white_check_mark: mgartner<br/>:white_check_mark: adityamaru<br/>:x: craig[bot]<br/><hr/>**craig[bot]** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=67266) it.</sub>

fqazi committed:
sql: use topological sorting inside new schema changer

Previously, the new schema changer attempted to use
a bubble sort which compared the nodes linked to ops
to generate an approximate order. Unfortunately, this
approach had its limitations when it came to elements
that had no edge linking them, which could lead to
incorrect order relative to dependency edges. To
address this, this patch moves the sorting to a simpler
topological sort which guarantees dependency order.

Release note: None

ajwerner commented on pkg/sql/schemachanger/scplan/plan.go:
```diff
@@ -223,132 +225,150 @@ func buildStages(init scpb.State, g *scgraph.Graph, params Params) []Stage {
-		curr := nodesToVisit[0]
-		if curr == target {
-			return true
+// SortOps sorts the operations into order based on
+// graph dependencies
+func SortOps(graph *scgraph.Graph, ops []scop.Op) {
```
I find myself surprised by this approach. I would have thought that it'd be easier to sort the edges than the ops. It seems to me like any time you need to sort, you already have the edges in scope. The edges then have the nodes and will generally avoid any need to go back and find nodes from ops.

ajwerner replied:
I believe that we can create a full topological sort of nodes at the outset of planning with the below code. We can then use that to sort op edges on their `To` node. I think that'll do it. Am I missing something? Consider the below code to generate a topological sort order:

```go
// rankNodes performs a topological sort of the nodes in the graph.
func rankNodes(g *Graph) map[*scpb.Node]int {
	backCycleExists := func(n *scpb.Node, de *DepEdge) bool {
		var foundBack bool
		_ = g.ForEachDepEdgeFrom(de.To(), func(maybeBack *DepEdge) error {
			foundBack = foundBack || maybeBack.To() == n
			return nil
		})
		return foundBack
	}
	l := list.New()
	marks := make(map[*scpb.Node]bool)
	var visit func(n *scpb.Node)
	visit = func(n *scpb.Node) {
		permanent, marked := marks[n]
		fmt.Println(marked, permanent, screl.NodeString(n))
		if marked && !permanent {
			panic("not a dag")
		}
		if marked && permanent {
			return
		}
		marks[n] = false
		g.ForEachDepEdgeFrom(n, func(de *DepEdge) error {
			// We want to eliminate cycles caused by swaps. In that
			// case, we want to pretend that there is no edge from the
			// add to the drop, and, in that way, the drop is ordered first.
			if n.Direction != scpb.Target_ADD || !backCycleExists(n, de) {
				visit(de.To())
			}
			return nil
		})
		if oe, ok := g.GetOpEdgeFrom(n); ok {
			visit(oe.To())
		}
		marks[n] = true
		l.PushFront(n)
	}

	_ = g.ForEachNode(func(n *scpb.Node) error {
		visit(n)
		return nil
	})
	rank := make(map[*scpb.Node]int, l.Len())
	for i, cur := 0, l.Front(); i < l.Len(); i++ {
		rank[cur.Value.(*scpb.Node)] = i
		cur = cur.Next()
	}
	return rank
}
```

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67266)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/scplan/plan.go, line 238 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-MfofaOwJQvUuGPOadft-r19-238:-MncFVDc0tuUqux2UBzD:bkzhibs) ([raw file](https://github.com/cockroachdb/cockroach/blob/26b1af827eead840f4acdcf3fc28a98ddff39f0c/pkg/sql/schemachanger/scplan/plan.go#L238)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I believe that we can create a full topological sort of nodes at the outset of planning with the below code. We can then use that to sort op edges on their `To` node. I think that'll do it. Am I missing something? Consider the below code to generate a topological sort order:

```go
// rankNodes performs a topological sort of the nodes in the graph.
func rankNodes(g *Graph) map[*scpb.Node]int {
	backCycleExists := func(n *scpb.Node, de *DepEdge) bool {
		var foundBack bool
		_ = g.ForEachDepEdgeFrom(de.To(), func(maybeBack *DepEdge) error {
			foundBack = foundBack || maybeBack.To() == n
			return nil
		})
		return foundBack
	}
	l := list.New()
	marks := make(map[*scpb.Node]bool)
	var visit func(n *scpb.Node)
	visit = func(n *scpb.Node) {
		permanent, marked := marks[n]
		fmt.Println(marked, permanent, screl.NodeString(n))
		if marked && !permanent {
			panic("not a dag")
		}
		if marked && permanent {
			return
		}
		marks[n] = false
		g.ForEachDepEdgeFrom(n, func(de *DepEdge) error {
			// We want to eliminate cycles caused by swaps. In that
			// case, we want to pretend that there is no edge from the
			// add to the drop, and, in that way, the drop is ordered first.
			if n.Direction != scpb.Target_ADD || !backCycleExists(n, de) {
				visit(de.To())
			}
			return nil
		})
		if oe, ok := g.GetOpEdgeFrom(n); ok {
			visit(oe.To())
		}
		marks[n] = true
		l.PushFront(n)
	}

	_ = g.ForEachNode(func(n *scpb.Node) error {
		visit(n)
		return nil
	})
	rank := make(map[*scpb.Node]int, l.Len())
	for i, cur := 0, l.Front(); i < l.Len(); i++ {
		rank[cur.Value.(*scpb.Node)] = i
		cur = cur.Next()
	}
	return rank
}
```
</blockquote></details>

Think I was way too focused on sorting the ops, so my approach was a bit heavy-handed. This algorithm looks good, and is much simpler, I'll replace the logic inside the graph.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/schemachanger/scbuild/table.go:
```diff
@@ -132,7 +194,7 @@ func (b *buildContext) alterTableAddColumn(
 			table, familyName, d.Family.Create, d.Family.IfNotExists,
 		)
-	} else {
+	} else if !d.IsVirtual() { // FIXME: Compute columns should not have families?
```
Stored ones definitely need families. Virtual ones I would think can live without them but I don't know whether or not they do.

ajwerner commented on pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go:
```diff
@@ -446,6 +448,19 @@ func (s *TestState) RemoveSyntheticDescriptor(id descpb.ID) {
+	allowedNewColumnNames []tree.Name,
+	allowImplicitPartitioning bool,
+) error {
+	return nil
```
ideally we'd track this and verify it in the tests

ajwerner commented on pkg/sql/schemachanger/scgraph/graph.go:
```diff
@@ -249,3 +265,50 @@ func (de *DepEdge) To() *scpb.Node { return de.to }
 // Name returns the name of the rule which generated this edge.
 func (de *DepEdge) Name() string { return de.rule }
+
+// GetNodeRanks fetches ranks of nodes in topological order
```
nit: period

ajwerner commented on pkg/sql/schemachanger/scjob/job.go:
```diff
@@ -66,6 +70,10 @@ func (n *newSchemaChangeResumer) Resume(ctx context.Context, execCtxI interface{
 		execCfg.Codec,
 		execCfg.Settings,
+		sql.MakeIndexValidator(execCfg.DB, execCfg.Codec, execCfg.InternalExecutor),
+		func(ctx context.Context, tableDesc *tabledesc.Mutable, indexDesc descpb.IndexDescriptor, partBy *tree.PartitionBy, allowedNewColumnNames []tree.Name, allowImplicitPartitioning bool) (newImplicitCols []catalog.Column, newPartitioning descpb.PartitioningDescriptor, err error) {
```
wrap this, or, perhaps better, write a function that has this signature. 

How would you feel about creating an interface as opposed to this function type? IDEs make it easy to implement the interface.

ajwerner commented on pkg/sql/schemachanger/scop/mutation.go:
```diff
@@ -119,6 +142,8 @@ type RemoveTypeBackRef struct {
 	TypeID descpb.ID
 }
 
+// FIXME: Check create index first
```
Can you add more to this?

ajwerner commented on pkg/sql/schemachanger/scop/mutation.go:
```diff
@@ -227,3 +252,19 @@ type RemoveSequenceOwnedBy struct {
+	RangePartitions []*scpb.RangePartitions
+}
+
+// NoOpInfo a no-op mutation operation.
```
Can you explain what this is for?

ajwerner commented on pkg/sql/schemachanger/scpb/scpb.proto:
```diff
@@ -189,3 +218,26 @@ message Database {
   repeated uint32 dependentObjects = 2  [(gogoproto.customname) = "DependentObjects", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.ID"];
 }
+
+// FIXME: Dead code
```
?

ajwerner commented on pkg/sql/schemachanger/scplan/plan.go:
```diff
@@ -151,13 +156,20 @@ func buildStages(init scpb.State, g *scgraph.Graph, params Params) []Stage {
 			isStageRevertible = revertible == 1
-			for i, ts := range cur {
-				for _, e := range edges {
+			// FIXME: Flip the loop
```
can you say more or do this?

ajwerner commented on pkg/sql/schemachanger/scplan/plan.go:
```diff
@@ -223,132 +235,10 @@ func buildStages(init scpb.State, g *scgraph.Graph, params Params) []Stage {
 		}
 		// Sort ops based on graph dependencies.
-		sortOps(g, s.Ops.Slice())
+		//SortOps(g, s.Ops.Slice())
```
detritus

ajwerner commented on pkg/sql/schemachanger/scplan/plan_test.go:
```diff
@@ -191,3 +193,84 @@ func marshalOps(t *testing.T, plan *scplan.Plan) string {
 	return stages.String()
 }
+
+// TestPlanGraphSort sanity checks sorting of the graph.
```
it seems like this test can go in `scgraph`. I wonder if we can rework it to be easier to read and extend.

ajwerner reviewed (commented):
My comments are mostly superficial and point towards tightening up the commentary.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67266)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/schemachanger/scjob/job.go, line 74 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mdw4TzBD5mU_aJpTEUj-r24-74:-Mo-EWxvDNaxzGez5S_W:b-oxsr0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264e8172499eb520f2cf9ebd4e545c41125a59d/pkg/sql/schemachanger/scjob/job.go#L74)):*
> How would you feel about creating an interface as opposed to this function type? IDEs make it easy to implement the interface.
___
*[pkg/sql/schemachanger/scop/mutation.go, line 145 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mdw4TzBD5mU_aJpTEUk-r24-145:-Mo-9nNO1uFpSbo0w2Ub:bf18xut) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264e8172499eb520f2cf9ebd4e545c41125a59d/pkg/sql/schemachanger/scop/mutation.go#L145)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you add more to this?
</blockquote></details>

Stale comment left over from when I started off.
___
*[pkg/sql/schemachanger/scop/mutation.go, line 267 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mdw4TzBD5mU_aJpTEUk-r24-267:-Mo-9tHG5dO5fXNj8c4O:bxge9qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264e8172499eb520f2cf9ebd4e545c41125a59d/pkg/sql/schemachanger/scop/mutation.go#L267)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you explain what this is for?
</blockquote></details>

// NoOpInfo a no-op mutation operation, which is allows
// an emit function to conditionally generate operations.
___
*[pkg/sql/schemachanger/scpb/scpb.proto, line 222 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mdw4TzBD5mU_aJpTEUp-r24-222:-Mo-9uJl92LoyM3twci2:bjpluo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264e8172499eb520f2cf9ebd4e545c41125a59d/pkg/sql/schemachanger/scpb/scpb.proto#L222)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

? 
</blockquote></details>

Stale comment from development, I should have checked my diffs for these.

Done.
___
*[pkg/sql/schemachanger/scplan/plan.go, line 170 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-MfofaOwJQvUuGPOadft-r24-170:-Mo-A-1W5_YGQx8mhSld:blkexc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264e8172499eb520f2cf9ebd4e545c41125a59d/pkg/sql/schemachanger/scplan/plan.go#L170)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you say more or do this?
</blockquote></details>

The edge and cur loops were flipped, which was a reminder for me.

Done.
___
*[pkg/sql/schemachanger/scplan/plan.go, line 238 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-MfofaOwJQvUuGPOadft-r24-238:-Mo-A9806UUXpVMZFGyg:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264e8172499eb520f2cf9ebd4e545c41125a59d/pkg/sql/schemachanger/scplan/plan.go#L238)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

detritus
</blockquote></details>

Done
___
*[pkg/sql/schemachanger/scplan/plan_test.go, line 197 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mdw4TzBD5mU_aJpTEUr-r24-197:-Mo-e_fRDF6J_Weq4wwW:b5oafav) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264e8172499eb520f2cf9ebd4e545c41125a59d/pkg/sql/schemachanger/scplan/plan_test.go#L197)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it seems like this test can go in `scgraph`. I wonder if we can rework it to be easier to read and extend. 
</blockquote></details>

Reworked this to have more syntatic sugar and more readability.

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> when it builds 

> Reviewed 1 of 58 files at r25, 1 of 33 files at r26, 4 of 46 files at r27.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-:-Mo-fuA67dtdSI1gpocG:b-c4sfxf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/schema_changer_ccl.go, line 50 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mo-fqMP4o0NRtnlpA0I:-Mo-fqMP4o0NRtnlpA0J:b-f4k1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4dbe1f6fac596aa3eac107688b91113677236c/pkg/sql/schema_changer_ccl.go#L50)):*
> ```Go
> 		allowImplicitPartitioning)
> }
> func MakeCCLCallbacks(
> ```

linter is going to want a comment here
___
*[pkg/sql/schemachanger/scexec/dependencies.go, line 99 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mo-fGpDDy21clZ7ft77:-Mo-fGpDDy21clZ7ft78:b-az4kd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4dbe1f6fac596aa3eac107688b91113677236c/pkg/sql/schemachanger/scexec/dependencies.go#L99)):*
> ```
> // CCLCallbacks provides an interface that implements CCL exclusive
> // callbacks.
> type CCLCallbacks interface {
> ```
nit: call this `Partitioner`? The fact that it's CCL isn't that important or meaningful. 
___
*[pkg/sql/schemachanger/scplan/plan.go, line 236 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/67266#-Mo-fbxBBYV64knBUgQm:-Mo-fbxBBYV64knBUgQn:b-lnhmax) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4dbe1f6fac596aa3eac107688b91113677236c/pkg/sql/schemachanger/scplan/plan.go#L236)):*
> ```Go
> 			break
> 		}
> 		// Sort ops based on graph dependencies.
> ```

This comment is now stale.


<!-- Sent from Reviewable.io -->

fqazi commented:
@ajwerner TFTR, thanks for the lightening faster review :)

fqazi commented:
bors r+

ajwerner commented:
Merge skew with the interleaved removal.

bors r-

fqazi commented:
bors r+
